torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,325 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+    weights.push_back(1);
+  }
+};
+
+//! Concretize broadcast axes, i.e. identifying a non-broadcast
+//! IterDomain that the broadcast IterDomain can map to.
+//!
+//! This traversal processes root domains only, concretization works by
+//! inspecting pointwise ops, e.g. : T2 [i0,i1] = T1[i0,B0] + T0[i0,i1]
+//! will concretize axis B0 to i1
+//!
+class ConcretizeDomain : public BackwardVisitor {
+ public:
+  //! traverses the graph backward from outputs
+  //! to identify all concretizing opportunities
+  //!
+  ConcretizeDomain(Fusion* fusion) {
+    traverseFrom(fusion, fusion->outputs(), false);
+  }
+
+  //! API call to run the concretize pass and return the
+  //! axis that bcast_dom concretizes to
+  //!
+  static const IterDomain* getConcreteDomain(IterDomain* bcast_dom) {
+    ConcretizeDomain cd(bcast_dom->fusion());
+
+    // remove this assertion once we support broadcast on output
+    TORCH_INTERNAL_ASSERT(cd.canConcretize(bcast_dom));
+    return cd.concretized(bcast_dom);
+  }
+
+  // returns true if either id is not a broadcast or
+  // the traversal has found a concretized axis for id
+  bool canConcretize(IterDomain* id) {
+    return !id->isBroadcast() || bcast_domain_map_.count(id);
+  }
+
+  // returns the concretized id recorded from traversal
+  IterDomain* concretized(IterDomain* id) {
+    TORCH_INTERNAL_ASSERT(canConcretize(id));
+    if (!id->isBroadcast())
+      return id;
+    return bcast_domain_map_.at(id);
+  }
+
+ private:
+  using MapType = std::unordered_map<IterDomain*, IterDomain*>;
+  MapType bcast_domain_map_;
+
+  // utility to inspect a pointwise operator and
+  // record concretize opportunities
+  void concretizePwOp(Expr* e);
+
+  // utility to record new concretize opportunity
+  void concretizeTo(IterDomain* id, IterDomain* To) {
+    TORCH_INTERNAL_ASSERT(id->isBroadcast() && !To->isBroadcast());
+    bcast_domain_map_[id] = concretized(To);
+  }
+
+  void handle(ReductionOp* rop) override {
+    concretizePwOp(rop->asExpr());
+  }
+
+  void handle(UnaryOp* uop) override {
+    concretizePwOp(uop->asExpr());
+  }
+
+  void handle(BinaryOp* bop) override {
+    concretizePwOp(bop->asExpr());
+  }
+
+  void handle(TernaryOp* top) override {
+    concretizePwOp(top->asExpr());
+  };
+};
+
+void ConcretizeDomain::concretizePwOp(Expr* e) {
+  TensorView* tv = *ir_utils::filterByType<TensorView>(e->outputs()).begin();
+
+  // won't use this function on a reduction op, so root domain only
+  std::vector<IterDomain*> io = tv->getRootDomain();
+
+  for (auto* i : ir_utils::filterByType<TensorView>(e->inputs())) {
+    std::vector<IterDomain*> ii =
+        TensorDomain::noReductions(i->getMaybeRFactorDomain());
+    TORCH_INTERNAL_ASSERT(ii.size() == io.size());
+
+    for (size_t it = 0; it < ii.size(); it++) {
+      if (!canConcretize(io[it]))
+        continue;
+
+      if (!canConcretize(ii[it]))
+        concretizeTo(ii[it], concretized(io[it]));
+    }
+  }
+} // ConcretizeDomain::pointWiseOpMap
```

Comment:
nit: no need for scope ending comment here

Reply:
updated. Thanks!